validator/core/src/output/data_type.rs

```diff
@@ -0,0 +1,356 @@
+//! Module for dealing with Substrait's type system.
+//!
+//! See [`DataType`].
+
+use crate::output::diagnostic;
+use crate::output::extension;
+use crate::output::primitive_data;
+use std::collections::HashSet;
+use std::sync::Arc;
+use strum_macros::{Display, EnumString};
+
+/// A Substrait data type.
+#[derive(Clone, Debug, PartialEq)]
+pub struct DataType {
+    /// Type class (simple, compound, or user-defined).
+    pub class: Class,
+
+    /// Nullability.
+    pub nullable: bool,
+
+    /// Type variation, if any.
+    pub variation: Option<Arc<extension::Reference<extension::TypeVariation>>>,
+
+    /// Type parameters for non-simple types.
+    pub parameters: Vec<Parameter>,
+}
+
+impl std::fmt::Display for DataType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.class)?;
+        if self.nullable {
+            write!(f, "?")?;
+        }
+        if let Some(variation) = &self.variation {
+            write!(f, "[{}]", variation)?;
+        }
+        if self.parameters.is_empty() {
```

Comment:
I guess this should be:
```suggestion
        if !self.parameters.is_empty() {
```